src/main.rs

```diff
@@ -0,0 +1,112 @@
+use std::fs::File;
+use std::io::Write;
+use std::sync::Arc;
+
+use clap::{App, Arg, ArgMatches, SubCommand};
+use http::{Request, Response};
+
+use third_wheel::*;
+
+#[tokio::main]
+async fn main() -> SafeResult {
+    let matches = App::new("third-wheel")
+        .version("0.1")
+        .author("Chris Campbell")
+        .about("A Rust clone of mitmproxy for fast and lightweight TLS proxying")
+        .subcommand(SubCommand::with_name("mitm")
+                    .about("Run a mitm proxy")
+                    .arg(Arg::with_name("port")
+                         .short("p")
+                         .help("Port to connect to")
+                         .required(false)
+                         .default_value("8080")
+                         .validator(|p| if let Err(_e) = p.parse::<u16>() {
+                             Err(String::from("Expected an integer"))
+                         } else { Ok(()) }
+                         )
+                    )
+        )
+        .subcommand(SubCommand::with_name("http-proxy")
+                    .about("Run a simple http proxy")
+                    .arg(Arg::with_name("port")
+                         .short("p")
+                         .help("Port to connect to")
+                         .required(false)
+                         .default_value("8080")
+                         .validator(|p| if let Err(_e) = p.parse::<u16>() {
+                             Err(String::from("Expected an integer"))
+                         } else { Ok(()) }
+                         )
+                    )
+        )
+        .subcommand(SubCommand::with_name("sign-cert-for-domain")
+                    .about("Sign a x509 certificate for a given domain")
+                    .arg(Arg::from_usage("<DOMAIN> 'The domain to sign the certificate for'"))
+                    .arg(Arg::from_usage("-o --outfile=[outfile] 'The file to store the certificate in'")
+                         .default_value("site.pem"))
+
+                    .arg(Arg::from_usage("-c --ca-cert-file=[cert_file] 'The pem file containing the ca certificate'")
+                         .default_value("./ca/ca_certs/cert.pem"))
+                    .arg(Arg::from_usage("-k --ca-key-file=[key_file] 'The pem file containing the ca key'")
+                         .default_value("./ca/ca_certs/key.pem"))
+        ).subcommand(SubCommand::with_name("testing"))
+        .get_matches();
+    run(matches).await
+}
+
+async fn run(matches: ArgMatches<'_>) -> SafeResult {
+    match matches.subcommand() {
+        ("testing", Some(_m)) => testing_main().await,
+        ("mitm", Some(m)) => {
+            struct EmptyCapturer;
```

Comment:
`#[derive(Clone, Copy)]`

Always derive Copy and Clone when possible. there is also a linter for that - https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/builtin/static.MISSING_COPY_IMPLEMENTATIONS.html